pkg/reader/usdc_reader.go

```diff
@@ -0,0 +1,282 @@
+	}
+
+	// Slice the first 32-byte segment
+	copy(result[:], m.Arg0[:32])
+
+	return result, nil
+}
+
+func NewUSDCMessageReader(
+	tokensConfig map[cciptypes.ChainSelector]pluginconfig.USDCCCTPTokenConfig,
+	contractReaders map[cciptypes.ChainSelector]contractreader.ContractReaderFacade,
+) (USDCMessageReader, error) {
+	boundContracts := make(map[cciptypes.ChainSelector]types.BoundContract)
+	for chainSelector, token := range tokensConfig {
+		bytesAddress, err := typconv.AddressStringToBytes(token.SourceMessageTransmitterAddr, uint64(chainSelector))
+		if err != nil {
+			return nil, err
+		}
+
+		contract, err := bindFacadeReaderContract(
+			context.Background(),
+			contractReaders,
+			chainSelector,
+			consts.ContractNameCCTPMessageTransmitter,
+			bytesAddress,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("unable to bind message transmitter for chain %d: %w", chainSelector, err)
+		}
+		boundContracts[chainSelector] = contract
+	}
+
+	return usdcMessageReader{
+		contractReaders: contractReaders,
+		cctpDestDomain:  CCTPDestDomains,
+		boundContracts:  boundContracts,
+	}, nil
+}
+
+func (u usdcMessageReader) MessageHashes(
+	ctx context.Context,
+	source, dest cciptypes.ChainSelector,
+	tokens map[exectypes.MessageTokenID]cciptypes.RampTokenAmount,
+) (map[exectypes.MessageTokenID]MessageHash, error) {
+	// 1. Extract 3rd word from the MessageSent(bytes) - it's going to be our identifier
+	eventIDs, err := u.recreateMessageTransmitterEvents(dest, tokens)
+	if err != nil {
+		return nil, err
+	}
+
+	// 2. Query the MessageTransmitter contract for the MessageSent events based on the 3rd words.
+	// We need entire MessageSent payload to use that with the Attestation API
+	cr, ok := u.boundContracts[source]
+	if !ok {
+		return nil, fmt.Errorf("no contract bound for chain %d", source)
+	}
+
+	iter, err := u.contractReaders[source].QueryKey(
+		ctx,
+		cr,
+		query.KeyFilter{
+			Key:         consts.EventNameCCTPMessageSent,
+			Expressions: []query.Expression{},
+		},
+		query.NewLimitAndSort(query.Limit{}, query.NewSortBySequence(query.Asc)),
+		&MessageSentEvent{},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error querying contract reader for chain %d: %w", source, err)
+	}
+
+	messageSentEvents := make(map[eventID]MessageHash)
+	for _, item := range iter {
+		event, ok1 := item.Data.(*MessageSentEvent)
+		if !ok1 {
+			return nil, fmt.Errorf("failed to cast %v to Message", item.Data)
+		}
+		e, err1 := event.unpackID()
+		if err1 != nil {
+			return nil, err1
+		}
+		messageSentEvents[e] = event.Arg0
+	}
+
+	// 3. This should be done by ChainReader - picking only events matching eventIDs.
+	// Right now ChainReader doesn't support filtering by specific data words
+	out := make(map[exectypes.MessageTokenID]MessageHash)
+	for tokenID, messageID := range eventIDs {
+		messageHash, ok1 := messageSentEvents[messageID]
+		if !ok1 {
+			// Token not available in the source chain
+			continue
+		}
+		out[tokenID] = messageHash
```

Comment:
double checking, the tokens will be always in same order? so out[0] will always mean the same token?

Reply:
Hmm, we do have a map here, so ordering doesn't really matter

Reply:
`tokenId` is a composite key as you suggested in a previous PR :P
```go
type MessageTokenID struct {
	SeqNr cciptypes.SeqNum
	Index int
}
```